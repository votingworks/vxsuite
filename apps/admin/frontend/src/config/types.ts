import {
  BallotStyleId,
  ContestId,
  PrecinctId,
  PromiseOr,
} from '@votingworks/types';
import type { ManualResultsVotingMethod } from '@votingworks/admin-backend';

// Events
export type InputEventFunction = (
  event: React.FormEvent<HTMLInputElement>
) => PromiseOr<void>;
export type TextareaEventFunction = (
  event: React.FormEvent<HTMLTextAreaElement>
) => PromiseOr<void>;

// Router Props
export interface ManualDataEntryScreenProps {
  precinctId: PrecinctId;
  ballotStyleId: BallotStyleId;
  votingMethod: ManualResultsVotingMethod;
}
export interface SmartcardsScreenProps {
  smartcardType: string;
}
export interface WriteInsAdjudicationScreenProps {
  contestId: ContestId;
}

export enum ResultsFileType {
  CastVoteRecord = 'cvr',
  All = 'all',
}

export interface CastVoteRecordFilePreprocessedData {
  readonly name: string;
  readonly path: string;
  readonly cvrCount: number;
  readonly scannerIds: readonly string[];
  readonly exportTimestamp: Date;
  readonly isTestModeResults: boolean;
}

export type Iso8601Timestamp = string;
