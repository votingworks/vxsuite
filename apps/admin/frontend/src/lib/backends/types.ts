import { FullElectionManualTally } from '@votingworks/types';

export interface AddCastVoteRecordFileResult {
  readonly wasExistingFile: boolean;
  readonly newlyAdded: number;
  readonly alreadyPresent: number;
}

/**
 * @deprecated these should be moved to `api.ts` as react-query hooks
 */
export interface ElectionManagerStoreBackend {
  /**
   * Loads the existing manual tallies.
   */
  loadFullElectionManualTally(): Promise<FullElectionManualTally | undefined>;

  /**
   * Updates the manual tally for a given source.
   */
  updateFullElectionManualTally(
    newFullElectionManualTally: FullElectionManualTally
  ): Promise<void>;

  /**
   * Removes the manual tally for a given source.
   */
  removeFullElectionManualTally(): Promise<void>;
}
