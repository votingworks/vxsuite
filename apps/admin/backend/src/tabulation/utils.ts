import { Tabulation } from '@votingworks/types';

/**
 * Tests whether CVR votes are empty.
 */
export function isBlankSheet(votes: Tabulation.Votes): boolean {
  return Object.values(votes).every((selections) => selections.length === 0);
}
