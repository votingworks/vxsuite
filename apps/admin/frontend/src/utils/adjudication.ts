import {
  BallotPageContestOptionLayout,
  ContestOptionId,
  Rect,
} from '@votingworks/types';

export function normalizeWriteInName(name: string): string {
  return name.toLowerCase().trim().replace(/\s+/, ' ');
}

export function getOptionCoordinates(
  optionsLayout: readonly BallotPageContestOptionLayout[],
  optionId: ContestOptionId
): Rect {
  const option = optionsLayout.find((opt) => opt.definition?.id === optionId);
  if (!option) {
    throw new Error('unable to find option in layout');
  }
  return option.bounds;
}
