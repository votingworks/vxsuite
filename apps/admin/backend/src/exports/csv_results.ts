// eslint-disable-next-line import/no-unresolved
import { stringify } from 'csv-stringify/sync';
import {
  writeInCandidate,
  Contest,
  Tabulation,
  electionHasPrimaryContest,
  ElectionDefinition,
  Id,
  AnyContest,
} from '@votingworks/types';
import { assert, assertDefined } from '@votingworks/basics';
import {
  convertGroupSpecifierToFilter,
  getBallotStyleById,
  getEmptyElectionResults,
  getPartyById,
  getPrecinctById,
  groupBySupportsZeroSplits,
  groupMapToGroupList,
  mergeFilters,
  populateSplits,
} from '@votingworks/utils';
import { Readable } from 'stream';
import { ScannerBatch } from '../types';
import { Store } from '../store';
import { tabulateElectionResults } from '../tabulation/full_results';

function generateHeaders({
  groupBy,
  isPrimaryElection,
}: {
  groupBy: Tabulation.GroupBy;
  isPrimaryElection: boolean;
}): string[] {
  const headers = [];

  if (groupBy.groupByVotingMethod) {
    headers.push('Voting Method');
  }

  if (groupBy.groupByPrecinct) {
    headers.push('Precinct');
    headers.push('Precinct ID');
  }

  if (
    isPrimaryElection &&
    (groupBy.groupByParty || groupBy.groupByBallotStyle)
  ) {
    headers.push('Party');
    headers.push('Party ID');
  }

  if (groupBy.groupByBallotStyle) {
    headers.push('Ballot Style ID');
  }

  if (groupBy.groupByScanner || groupBy.groupByBatch) {
    headers.push('Scanner ID');
  }

  if (groupBy.groupByBatch) {
    headers.push('Batch ID');
  }

  headers.push('Contest', 'Contest ID', 'Selection', 'Selection ID', 'Votes');

  return headers;
}

const VOTING_METHOD_LABEL: Record<Tabulation.VotingMethod, string> = {
  absentee: 'Absentee',
  precinct: 'Precinct',
  provisional: 'Provisional',
};

type BatchLookup = Record<string, ScannerBatch>;

function buildCsvRow({
  groupBy,
  groupSpecifier,
  electionDefinition,
  isPrimaryElection,
  batchLookup,
  contest,
  selection,
  selectionId,
  votes,
}: {
  groupBy: Tabulation.GroupBy;
  groupSpecifier: Tabulation.GroupSpecifier;
  electionDefinition: ElectionDefinition;
  isPrimaryElection: boolean;
  batchLookup: BatchLookup;
  contest: Contest;
  selection: string;
  selectionId: string;
  votes: number;
}): string {
  const values: string[] = [];

  if (groupBy.groupByVotingMethod) {
    assert(groupSpecifier.votingMethod !== undefined);
    values.push(VOTING_METHOD_LABEL[groupSpecifier.votingMethod]);
  }

  if (groupBy.groupByPrecinct) {
    assert(groupSpecifier.precinctId !== undefined);
    values.push(
      getPrecinctById(electionDefinition, groupSpecifier.precinctId).name
    );
    values.push(groupSpecifier.precinctId);
  }

  if (
    isPrimaryElection &&
    (groupBy.groupByParty || groupBy.groupByBallotStyle)
  ) {
    const partyId = (() => {
      if (groupBy.groupByParty) {
        return assertDefined(groupSpecifier.partyId);
      }

      return assertDefined(
        getBallotStyleById(
          electionDefinition,
          assertDefined(groupSpecifier.ballotStyleId)
        ).partyId
      );
    })();

    values.push(assertDefined(getPartyById(electionDefinition, partyId).name));
    values.push(partyId);
  }

  if (groupBy.groupByBallotStyle) {
    assert(groupSpecifier.ballotStyleId !== undefined);
    values.push(groupSpecifier.ballotStyleId);
  }

  if (groupBy.groupByScanner || groupBy.groupByBatch) {
    const scannerId = (() => {
      if (groupBy.groupByScanner) {
        return assertDefined(groupSpecifier.scannerId);
      }

      return assertDefined(batchLookup[assertDefined(groupSpecifier.batchId)])
        .scannerId;
    })();
    values.push(scannerId);
  }

  if (groupBy.groupByBatch) {
    values.push(assertDefined(groupSpecifier.batchId));
  }

  values.push(
    contest.title,
    contest.id,
    selection,
    selectionId,
    votes.toString()
  );

  return stringify([values]);
}

function generateBatchLookup(store: Store, electionId: Id): BatchLookup {
  const batches = store.getScannerBatches(electionId);
  const lookup: BatchLookup = {};
  for (const batch of batches) {
    lookup[batch.batchId] = batch;
  }
  return lookup;
}

function* generateRows({
  electionId,
  electionDefinition,
  groupBy,
  resultSplits,
  store,
  overallReportFilter,
}: {
  electionId: Id;
  electionDefinition: ElectionDefinition;
  groupBy: Tabulation.GroupBy;
  resultSplits: Tabulation.ElectionResultsGroupList;
  store: Store;
  overallReportFilter: Tabulation.Filter;
}): Generator<string> {
  const { election } = electionDefinition;
  const isPrimaryElection = electionHasPrimaryContest(election);
  const writeInCandidates = store.getWriteInCandidates({ electionId });
  const batchLookup = generateBatchLookup(store, assertDefined(electionId));

  for (const resultsSplit of resultSplits) {
    const effectiveFilter = mergeFilters(
      overallReportFilter,
      convertGroupSpecifierToFilter(resultsSplit)
    );
    const contestIds = new Set(
      store.getFilteredContests({
        electionId,
        filter: effectiveFilter,
      })
    );
    const includedContests: AnyContest[] = [];
    for (const contest of election.contests) {
      if (contestIds.has(contest.id)) {
        includedContests.push(contest);
      }
    }

    for (const contest of includedContests) {
      const contestWriteInCandidates = writeInCandidates.filter(
        (c) => c.contestId === contest.id
      );
      const contestResults = resultsSplit.contestResults[contest.id];
      assert(contestResults !== undefined);

      if (contest.type === 'candidate') {
        assert(contestResults.contestType === 'candidate');

        // official candidate rows
        for (const candidate of contest.candidates) {
          /* c8 ignore next -- trivial fallthrough zero branch */
          const votes = contestResults.tallies[candidate.id]?.tally ?? 0;
          yield buildCsvRow({
            groupBy,
            groupSpecifier: resultsSplit,
            electionDefinition,
            isPrimaryElection,
            batchLookup,
            contest,
            selection: candidate.name,
            selectionId: candidate.id,
            votes,
          });
        }

        // generic write-in row
        if (contest.allowWriteIns) {
          const votes = contestResults.tallies[writeInCandidate.id]?.tally ?? 0;
          if (votes) {
            yield buildCsvRow({
              groupBy,
              groupSpecifier: resultsSplit,
              electionDefinition,
              isPrimaryElection,
              batchLookup,
              contest,
              selection: writeInCandidate.name,
              selectionId: writeInCandidate.id,
              votes,
            });
          }
        }

        // adjudicated write-in rows
        for (const contestWriteInCandidate of contestWriteInCandidates) {
          /* c8 ignore next 2 -- trivial fallthrough zero branch */
          const votes =
            contestResults.tallies[contestWriteInCandidate.id]?.tally ?? 0;

          if (votes) {
            yield buildCsvRow({
              groupBy,
              groupSpecifier: resultsSplit,
              electionDefinition,
              isPrimaryElection,
              batchLookup,
              contest,
              selection: contestWriteInCandidate.name,
              selectionId: contestWriteInCandidate.id,
              votes,
            });
          }
        }
      } else if (contest.type === 'yesno') {
        assert(contestResults.contestType === 'yesno');
        yield buildCsvRow({
          groupBy,
          groupSpecifier: resultsSplit,
          electionDefinition,
          isPrimaryElection,
          batchLookup,
          contest,
          selection: 'Yes',
          selectionId: contest.yesOption?.id || 'yes',
          votes: contestResults.yesTally,
        });
        yield buildCsvRow({
          groupBy,
          groupSpecifier: resultsSplit,
          electionDefinition,
          isPrimaryElection,
          batchLookup,
          contest,
          selection: 'No',
          selectionId: contest.noOption?.id || 'no',
          votes: contestResults.noTally,
        });
      }

      yield buildCsvRow({
        groupBy,
        groupSpecifier: resultsSplit,
        electionDefinition,
        isPrimaryElection,
        batchLookup,
        contest,
        selection: 'Overvotes',
        selectionId: 'overvotes',
        votes: contestResults.overvotes,
      });

      yield buildCsvRow({
        groupBy,
        groupSpecifier: resultsSplit,
        electionDefinition,
        isPrimaryElection,
        batchLookup,
        contest,
        selection: 'Undervotes',
        selectionId: 'undervotes',
        votes: contestResults.undervotes,
      });
    }
  }
}

/**
 * Converts a tally for an election to a CSV file (represented as a string) of tally
 * results. Results are split according to the `groupBy` parameter. For each
 * additional split, one or more split metadata columns are added to each row.
 *  - `groupByBallotStyle` adds "Ballot Style ID" and, if a primary, "Party" and "Party ID"
 *  - `groupByParty` adds "Party" and "Party ID", if a primary
 *  - `groupByScanner` adds "Scanner ID"
 *  - `groupByBatch` adds "Batch ID" and "Scanner ID"
 *  - `groupByPrecinct` adds "Precinct" and "Precinct ID"
 *  - `groupByVotingMethod` adds "Voting Method"
 *
 * Notice that we are adding not only the metadata for the specified group but
 * metadata from inferable groups too. E.g., if we group by batch then we know
 * the scanner ID of each group.
 *
 * Returns the file as a `NodeJS.ReadableStream` emitting line by line.
 */
export async function generateResultsCsv({
  store,
  filter = {},
  groupBy = {},
}: {
  store: Store;
  filter?: Tabulation.Filter;
  groupBy?: Tabulation.GroupBy;
}): Promise<NodeJS.ReadableStream> {
  const electionId = store.getCurrentElectionId();
  assert(electionId !== undefined);
  const { electionDefinition } = assertDefined(store.getElection(electionId));
  const { election } = electionDefinition;

  const isPrimaryElection = electionHasPrimaryContest(election);
  const headerRow = stringify([
    generateHeaders({ groupBy, isPrimaryElection }),
  ]);

  const nonEmptySplits = await tabulateElectionResults({
    electionId,
    store,
    filter,
    groupBy,
    includeManualResults: true,
    includeWriteInAdjudicationResults: true,
  });

  const resultSplits = (() => {
    if (!groupBySupportsZeroSplits(groupBy)) {
      return groupMapToGroupList(nonEmptySplits);
    }

    const expectedSplits = store.getTabulationGroups({
      electionId,
      groupBy,
      filter,
    });

    return populateSplits({
      expectedSplits,
      nonEmptySplits,
      groupBy,
      makeEmptySplit: () => getEmptyElectionResults(election),
    });
  })();

  function* generateResultsCsvRows() {
    yield headerRow;

    for (const dataRow of generateRows({
      resultSplits,
      groupBy,
      electionDefinition,
      electionId: assertDefined(electionId),
      store,
      overallReportFilter: filter,
    })) {
      yield dataRow;
    }
  }

  return Readable.from(generateResultsCsvRows());
}
