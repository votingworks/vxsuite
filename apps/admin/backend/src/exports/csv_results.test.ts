import { electionMinimalExhaustiveSampleFixtures } from '@votingworks/fixtures';
import { Tabulation } from '@votingworks/types';
import { Buffer } from 'buffer';
import { buildElectionResultsFixture } from '@votingworks/utils';
import { generateResultsCsv } from './csv_results';
import { WriteInCandidateRecord } from '../types';

test('generateResultsCsv', async () => {
  const { election } =
    electionMinimalExhaustiveSampleFixtures.electionDefinition;
  const electionResultsByPrecinctAndVotingMethod: Tabulation.GroupedElectionResults =
    {
      'root&precinct-1&absentee': {
        precinctId: 'precinct-1',
        votingMethod: 'absentee',
        ...buildElectionResultsFixture({
          election,
          includeGenericWriteIn: true,
          cardCounts: {
            hmpb: [10],
            bmd: 0,
          },
          contestResultsSummaries: {
            'zoo-council-mammal': {
              type: 'candidate',
              ballots: 10,
              overvotes: 3,
              undervotes: 2,
              officialOptionTallies: {
                lion: 20,
                'write-in': 3,
              },
              writeInOptionTallies: {
                unofficial: {
                  name: 'Unofficial Write-In',
                  tally: 2,
                },
              },
            },
          },
        }),
      },
      'root&precinct-2&precinct': {
        precinctId: 'precinct-2',
        votingMethod: 'precinct',
        ...buildElectionResultsFixture({
          election,
          includeGenericWriteIn: true,
          cardCounts: {
            hmpb: [30, 0],
            bmd: 0,
          },
          contestResultsSummaries: {
            fishing: {
              type: 'yesno',
              ballots: 10,
              overvotes: 3,
              undervotes: 2,
              yesTally: 1,
              noTally: 4,
            },
          },
        }),
      },
    };

  const writeInCandidates: WriteInCandidateRecord[] = [
    {
      electionId: 'id',
      id: 'unofficial',
      name: 'Unofficial Write-In',
      contestId: 'zoo-council-mammal',
    },
  ];

  const resultsGenerator = generateResultsCsv({
    electionResultsByPrecinctAndVotingMethod,
    election,
    writeInCandidates,
  });

  const chunks = [];

  for await (const chunk of resultsGenerator) {
    chunks.push(Buffer.from(chunk));
  }

  const csvString = Buffer.concat(chunks).toString('utf-8');

  expect(csvString).toMatchSnapshot();
});
