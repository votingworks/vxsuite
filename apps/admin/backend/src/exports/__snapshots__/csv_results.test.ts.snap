// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateResultsCsv 1`] = `
"Contest,Contest ID,Selection,Selection ID,Precinct,Precinct ID,Voting Method,Votes
Best Animal,best-animal-mammal,Horse,horse,Precinct 1,precinct-1,Absentee,0
Best Animal,best-animal-mammal,Otter,otter,Precinct 1,precinct-1,Absentee,0
Best Animal,best-animal-mammal,Fox,fox,Precinct 1,precinct-1,Absentee,0
Best Animal,best-animal-mammal,Overvotes,,Precinct 1,precinct-1,Absentee,0
Best Animal,best-animal-mammal,Undervotes,,Precinct 1,precinct-1,Absentee,0
Best Animal,best-animal-fish,Seahorse,seahorse,Precinct 1,precinct-1,Absentee,0
Best Animal,best-animal-fish,Salmon,salmon,Precinct 1,precinct-1,Absentee,0
Best Animal,best-animal-fish,Overvotes,,Precinct 1,precinct-1,Absentee,0
Best Animal,best-animal-fish,Undervotes,,Precinct 1,precinct-1,Absentee,0
Zoo Council,zoo-council-mammal,Zebra,zebra,Precinct 1,precinct-1,Absentee,0
Zoo Council,zoo-council-mammal,Lion,lion,Precinct 1,precinct-1,Absentee,20
Zoo Council,zoo-council-mammal,Kangaroo,kangaroo,Precinct 1,precinct-1,Absentee,0
Zoo Council,zoo-council-mammal,Elephant,elephant,Precinct 1,precinct-1,Absentee,0
Zoo Council,zoo-council-mammal,Write-In,write-in,Precinct 1,precinct-1,Absentee,3
Zoo Council,zoo-council-mammal,Unofficial Write-In,unofficial,Precinct 1,precinct-1,Absentee,2
Zoo Council,zoo-council-mammal,Overvotes,,Precinct 1,precinct-1,Absentee,3
Zoo Council,zoo-council-mammal,Undervotes,,Precinct 1,precinct-1,Absentee,2
Zoo Council,aquarium-council-fish,Manta Ray,manta-ray,Precinct 1,precinct-1,Absentee,0
Zoo Council,aquarium-council-fish,Pufferfish,pufferfish,Precinct 1,precinct-1,Absentee,0
Zoo Council,aquarium-council-fish,Rockfish,rockfish,Precinct 1,precinct-1,Absentee,0
Zoo Council,aquarium-council-fish,Triggerfish,triggerfish,Precinct 1,precinct-1,Absentee,0
Zoo Council,aquarium-council-fish,Overvotes,,Precinct 1,precinct-1,Absentee,0
Zoo Council,aquarium-council-fish,Undervotes,,Precinct 1,precinct-1,Absentee,0
Ballot Measure 1,new-zoo-either,Yes,new-zoo-either-approved,Precinct 1,precinct-1,Absentee,0
Ballot Measure 1,new-zoo-either,No,new-zoo-neither-approved,Precinct 1,precinct-1,Absentee,0
Ballot Measure 1,new-zoo-either,Overvotes,,Precinct 1,precinct-1,Absentee,0
Ballot Measure 1,new-zoo-either,Undervotes,,Precinct 1,precinct-1,Absentee,0
Ballot Measure 1,new-zoo-pick,Yes,new-zoo-safari,Precinct 1,precinct-1,Absentee,0
Ballot Measure 1,new-zoo-pick,No,new-zoo-traditional,Precinct 1,precinct-1,Absentee,0
Ballot Measure 1,new-zoo-pick,Overvotes,,Precinct 1,precinct-1,Absentee,0
Ballot Measure 1,new-zoo-pick,Undervotes,,Precinct 1,precinct-1,Absentee,0
Ballot Measure 3,fishing,Yes,ban-fishing,Precinct 1,precinct-1,Absentee,0
Ballot Measure 3,fishing,No,allow-fishing,Precinct 1,precinct-1,Absentee,0
Ballot Measure 3,fishing,Overvotes,,Precinct 1,precinct-1,Absentee,0
Ballot Measure 3,fishing,Undervotes,,Precinct 1,precinct-1,Absentee,0
Best Animal,best-animal-mammal,Horse,horse,Precinct 1,precinct-1,Precinct,0
Best Animal,best-animal-mammal,Otter,otter,Precinct 1,precinct-1,Precinct,0
Best Animal,best-animal-mammal,Fox,fox,Precinct 1,precinct-1,Precinct,0
Best Animal,best-animal-mammal,Overvotes,,Precinct 1,precinct-1,Precinct,0
Best Animal,best-animal-mammal,Undervotes,,Precinct 1,precinct-1,Precinct,0
Best Animal,best-animal-fish,Seahorse,seahorse,Precinct 1,precinct-1,Precinct,0
Best Animal,best-animal-fish,Salmon,salmon,Precinct 1,precinct-1,Precinct,0
Best Animal,best-animal-fish,Overvotes,,Precinct 1,precinct-1,Precinct,0
Best Animal,best-animal-fish,Undervotes,,Precinct 1,precinct-1,Precinct,0
Zoo Council,zoo-council-mammal,Zebra,zebra,Precinct 1,precinct-1,Precinct,0
Zoo Council,zoo-council-mammal,Lion,lion,Precinct 1,precinct-1,Precinct,0
Zoo Council,zoo-council-mammal,Kangaroo,kangaroo,Precinct 1,precinct-1,Precinct,0
Zoo Council,zoo-council-mammal,Elephant,elephant,Precinct 1,precinct-1,Precinct,0
Zoo Council,zoo-council-mammal,Overvotes,,Precinct 1,precinct-1,Precinct,0
Zoo Council,zoo-council-mammal,Undervotes,,Precinct 1,precinct-1,Precinct,0
Zoo Council,aquarium-council-fish,Manta Ray,manta-ray,Precinct 1,precinct-1,Precinct,0
Zoo Council,aquarium-council-fish,Pufferfish,pufferfish,Precinct 1,precinct-1,Precinct,0
Zoo Council,aquarium-council-fish,Rockfish,rockfish,Precinct 1,precinct-1,Precinct,0
Zoo Council,aquarium-council-fish,Triggerfish,triggerfish,Precinct 1,precinct-1,Precinct,0
Zoo Council,aquarium-council-fish,Overvotes,,Precinct 1,precinct-1,Precinct,0
Zoo Council,aquarium-council-fish,Undervotes,,Precinct 1,precinct-1,Precinct,0
Ballot Measure 1,new-zoo-either,Yes,new-zoo-either-approved,Precinct 1,precinct-1,Precinct,0
Ballot Measure 1,new-zoo-either,No,new-zoo-neither-approved,Precinct 1,precinct-1,Precinct,0
Ballot Measure 1,new-zoo-either,Overvotes,,Precinct 1,precinct-1,Precinct,0
Ballot Measure 1,new-zoo-either,Undervotes,,Precinct 1,precinct-1,Precinct,0
Ballot Measure 1,new-zoo-pick,Yes,new-zoo-safari,Precinct 1,precinct-1,Precinct,0
Ballot Measure 1,new-zoo-pick,No,new-zoo-traditional,Precinct 1,precinct-1,Precinct,0
Ballot Measure 1,new-zoo-pick,Overvotes,,Precinct 1,precinct-1,Precinct,0
Ballot Measure 1,new-zoo-pick,Undervotes,,Precinct 1,precinct-1,Precinct,0
Ballot Measure 3,fishing,Yes,ban-fishing,Precinct 1,precinct-1,Precinct,0
Ballot Measure 3,fishing,No,allow-fishing,Precinct 1,precinct-1,Precinct,0
Ballot Measure 3,fishing,Overvotes,,Precinct 1,precinct-1,Precinct,0
Ballot Measure 3,fishing,Undervotes,,Precinct 1,precinct-1,Precinct,0
Best Animal,best-animal-mammal,Horse,horse,Precinct 2,precinct-2,Absentee,0
Best Animal,best-animal-mammal,Otter,otter,Precinct 2,precinct-2,Absentee,0
Best Animal,best-animal-mammal,Fox,fox,Precinct 2,precinct-2,Absentee,0
Best Animal,best-animal-mammal,Overvotes,,Precinct 2,precinct-2,Absentee,0
Best Animal,best-animal-mammal,Undervotes,,Precinct 2,precinct-2,Absentee,0
Best Animal,best-animal-fish,Seahorse,seahorse,Precinct 2,precinct-2,Absentee,0
Best Animal,best-animal-fish,Salmon,salmon,Precinct 2,precinct-2,Absentee,0
Best Animal,best-animal-fish,Overvotes,,Precinct 2,precinct-2,Absentee,0
Best Animal,best-animal-fish,Undervotes,,Precinct 2,precinct-2,Absentee,0
Zoo Council,zoo-council-mammal,Zebra,zebra,Precinct 2,precinct-2,Absentee,0
Zoo Council,zoo-council-mammal,Lion,lion,Precinct 2,precinct-2,Absentee,0
Zoo Council,zoo-council-mammal,Kangaroo,kangaroo,Precinct 2,precinct-2,Absentee,0
Zoo Council,zoo-council-mammal,Elephant,elephant,Precinct 2,precinct-2,Absentee,0
Zoo Council,zoo-council-mammal,Overvotes,,Precinct 2,precinct-2,Absentee,0
Zoo Council,zoo-council-mammal,Undervotes,,Precinct 2,precinct-2,Absentee,0
Zoo Council,aquarium-council-fish,Manta Ray,manta-ray,Precinct 2,precinct-2,Absentee,0
Zoo Council,aquarium-council-fish,Pufferfish,pufferfish,Precinct 2,precinct-2,Absentee,0
Zoo Council,aquarium-council-fish,Rockfish,rockfish,Precinct 2,precinct-2,Absentee,0
Zoo Council,aquarium-council-fish,Triggerfish,triggerfish,Precinct 2,precinct-2,Absentee,0
Zoo Council,aquarium-council-fish,Overvotes,,Precinct 2,precinct-2,Absentee,0
Zoo Council,aquarium-council-fish,Undervotes,,Precinct 2,precinct-2,Absentee,0
Ballot Measure 1,new-zoo-either,Yes,new-zoo-either-approved,Precinct 2,precinct-2,Absentee,0
Ballot Measure 1,new-zoo-either,No,new-zoo-neither-approved,Precinct 2,precinct-2,Absentee,0
Ballot Measure 1,new-zoo-either,Overvotes,,Precinct 2,precinct-2,Absentee,0
Ballot Measure 1,new-zoo-either,Undervotes,,Precinct 2,precinct-2,Absentee,0
Ballot Measure 1,new-zoo-pick,Yes,new-zoo-safari,Precinct 2,precinct-2,Absentee,0
Ballot Measure 1,new-zoo-pick,No,new-zoo-traditional,Precinct 2,precinct-2,Absentee,0
Ballot Measure 1,new-zoo-pick,Overvotes,,Precinct 2,precinct-2,Absentee,0
Ballot Measure 1,new-zoo-pick,Undervotes,,Precinct 2,precinct-2,Absentee,0
Ballot Measure 3,fishing,Yes,ban-fishing,Precinct 2,precinct-2,Absentee,0
Ballot Measure 3,fishing,No,allow-fishing,Precinct 2,precinct-2,Absentee,0
Ballot Measure 3,fishing,Overvotes,,Precinct 2,precinct-2,Absentee,0
Ballot Measure 3,fishing,Undervotes,,Precinct 2,precinct-2,Absentee,0
Best Animal,best-animal-mammal,Horse,horse,Precinct 2,precinct-2,Precinct,0
Best Animal,best-animal-mammal,Otter,otter,Precinct 2,precinct-2,Precinct,0
Best Animal,best-animal-mammal,Fox,fox,Precinct 2,precinct-2,Precinct,0
Best Animal,best-animal-mammal,Overvotes,,Precinct 2,precinct-2,Precinct,0
Best Animal,best-animal-mammal,Undervotes,,Precinct 2,precinct-2,Precinct,0
Best Animal,best-animal-fish,Seahorse,seahorse,Precinct 2,precinct-2,Precinct,0
Best Animal,best-animal-fish,Salmon,salmon,Precinct 2,precinct-2,Precinct,0
Best Animal,best-animal-fish,Overvotes,,Precinct 2,precinct-2,Precinct,0
Best Animal,best-animal-fish,Undervotes,,Precinct 2,precinct-2,Precinct,0
Zoo Council,zoo-council-mammal,Zebra,zebra,Precinct 2,precinct-2,Precinct,0
Zoo Council,zoo-council-mammal,Lion,lion,Precinct 2,precinct-2,Precinct,0
Zoo Council,zoo-council-mammal,Kangaroo,kangaroo,Precinct 2,precinct-2,Precinct,0
Zoo Council,zoo-council-mammal,Elephant,elephant,Precinct 2,precinct-2,Precinct,0
Zoo Council,zoo-council-mammal,Overvotes,,Precinct 2,precinct-2,Precinct,0
Zoo Council,zoo-council-mammal,Undervotes,,Precinct 2,precinct-2,Precinct,0
Zoo Council,aquarium-council-fish,Manta Ray,manta-ray,Precinct 2,precinct-2,Precinct,0
Zoo Council,aquarium-council-fish,Pufferfish,pufferfish,Precinct 2,precinct-2,Precinct,0
Zoo Council,aquarium-council-fish,Rockfish,rockfish,Precinct 2,precinct-2,Precinct,0
Zoo Council,aquarium-council-fish,Triggerfish,triggerfish,Precinct 2,precinct-2,Precinct,0
Zoo Council,aquarium-council-fish,Overvotes,,Precinct 2,precinct-2,Precinct,0
Zoo Council,aquarium-council-fish,Undervotes,,Precinct 2,precinct-2,Precinct,0
Ballot Measure 1,new-zoo-either,Yes,new-zoo-either-approved,Precinct 2,precinct-2,Precinct,0
Ballot Measure 1,new-zoo-either,No,new-zoo-neither-approved,Precinct 2,precinct-2,Precinct,0
Ballot Measure 1,new-zoo-either,Overvotes,,Precinct 2,precinct-2,Precinct,0
Ballot Measure 1,new-zoo-either,Undervotes,,Precinct 2,precinct-2,Precinct,0
Ballot Measure 1,new-zoo-pick,Yes,new-zoo-safari,Precinct 2,precinct-2,Precinct,0
Ballot Measure 1,new-zoo-pick,No,new-zoo-traditional,Precinct 2,precinct-2,Precinct,0
Ballot Measure 1,new-zoo-pick,Overvotes,,Precinct 2,precinct-2,Precinct,0
Ballot Measure 1,new-zoo-pick,Undervotes,,Precinct 2,precinct-2,Precinct,0
Ballot Measure 3,fishing,Yes,ban-fishing,Precinct 2,precinct-2,Precinct,1
Ballot Measure 3,fishing,No,allow-fishing,Precinct 2,precinct-2,Precinct,4
Ballot Measure 3,fishing,Overvotes,,Precinct 2,precinct-2,Precinct,3
Ballot Measure 3,fishing,Undervotes,,Precinct 2,precinct-2,Precinct,2
"
`;
